File: linked-lists/copy-list-with-random-pointer/copy-list-with-random-pointer.java
class Solution {
  public Node copyRandomList(Node head) {

    if (head == null)
      return null;
    Node curr = head;

    while (curr != null) {
      Node temp = curr.next;
      curr.next = new Node(curr.val, temp, null);
      curr = temp;
    }

    curr = head;

    while (curr != null) {
      if (curr.random != null) {
        curr.next.random = curr.random.next;
      }
      curr = curr.next.next;
    }

    curr = head;
    Node copy = head.next;

    while (curr != null) {
      Node next = curr.next.next;
      Node c = curr.next;
      curr.next = next;

      if (next != null) {
        c.next = next.next;
      }
      curr = next;
    }

    return copy;

  }
}

File: linked-lists/copy-list-with-random-pointer/copy-list-with-random-pointer.go
package copylistwithrandompointer

type Node struct {
	Val    int
	Next   *Node
	Random *Node
}

func copyRandomList(head *Node) *Node {
	if head == nil {
		return nil
	}

	curr := head

	for curr != nil {
		temp := curr.Next
		curr.Next = &Node{Val: curr.Val, Next: temp, Random: nil}
		curr = temp
	}

	curr = head

	for curr != nil {
		if curr.Random != nil {
			curr.Next.Random = curr.Random.Next
		}
		curr = curr.Next.Next
	}

	curr = head
	headCopy := head.Next

	for curr != nil {
		next := curr.Next.Next
		c := curr.Next
		curr.Next = next

		if next != nil {
			c.Next = next.Next
		}

		curr = next
	}

	return headCopy
}
